lib/puppet/provider/package/dpkg.rb

```diff
@@ -44,16 +43,23 @@ def self.instances
   # Note: self:: is required here to keep these constants in the context of what will
   # eventually become this Puppet::Type::Package::ProviderDpkg class.
   self::DPKG_QUERY_FORMAT_STRING = %Q{'${Status} ${Package} ${Version}\\n'}
+  self::DPKG_QUERY_PROVIDES_FORMAT_STRING = %Q{'${Status} ${Package} ${Version} [${Provides}]\\n'}
   self::FIELDS_REGEX = %r{^(\S+) +(\S+) +(\S+) (\S+) (\S*)$}
+  self::FIELDS_REGEX_WITH_PROVIDES = %r{^(\S+) +(\S+) +(\S+) (\S+) (\S*) \[.*\]$}
   self::FIELDS= [:desired, :error, :status, :name, :ensure]
 
+  def self.defaultto_allow_virtual
+    false
+  end
```

Comment:
`defaultto_allow_virtual` needs to be an instance method so that `if provider.respond_to?(:defaultto_allow_virtual)` works. We need a test to make sure the default value for `allow_virtual` for dpkg and apt is still false.

Reply:
added test: https://github.com/puppetlabs/puppet/pull/7701/files#diff-161079850524215fce04feb6db799d41R377

Reply:
added test here: https://github.com/puppetlabs/puppet/pull/7701/files#diff-8497ab1e5ae8249a6f30a07bb8bf91fdR550